extensions/azure/assetindex-cosmos/src/main/java/org/eclipse/dataspaceconnector/assetindex/azure/CosmosAssetIndexExtension.java

```diff
@@ -0,0 +1,119 @@
+import org.eclipse.dataspaceconnector.common.string.StringUtils;
+import org.eclipse.dataspaceconnector.cosmos.azure.AbstractCosmosConfig;
+import org.eclipse.dataspaceconnector.spi.EdcException;
+import org.eclipse.dataspaceconnector.spi.asset.AssetIndex;
+import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+import org.eclipse.dataspaceconnector.spi.security.Vault;
+import org.eclipse.dataspaceconnector.spi.system.ServiceExtension;
+import org.eclipse.dataspaceconnector.spi.system.ServiceExtensionContext;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Provides a persistent implementation of the {@link org.eclipse.dataspaceconnector.spi.asset.AssetIndex} using CosmosDB.
+ */
+public class CosmosAssetIndexExtension implements ServiceExtension {
+
+    private static final String NAME = "CosmosDB Asset Index";
+
+    private Monitor monitor;
+
+    @Override
+    public Set<String> provides() {
+        return Set.of(AssetIndex.FEATURE);
+    }
+
+    @Override
+    public void initialize(ServiceExtensionContext context) {
+        monitor = context.getMonitor();
+        monitor.info(String.format("Initializing %s extension...", NAME));
+
+        var configuration = new AssetIndexCosmosConfig(context);
+        Vault service = context.getService(Vault.class);
+
+        // create client
+        var cosmosClient = createClient(service, configuration.getAccountName(), Collections.singletonList(configuration.getPreferredRegion()));
+
+        // get database, throw exception if not exists
+        var database = createDatabase(configuration, cosmosClient);
+        var container = getContainer(database, configuration);
+
+        context.registerService(AssetIndex.class, new CosmosAssetIndex(
+                container, configuration.getPartitionKey(), context.getTypeManager(), context.getService(RetryPolicy.class), configuration.isQueryMetricsEnabled()));
+
+        context.getTypeManager().registerTypes(AssetDocument.class);
+        monitor.info(String.format("Initialized %s extension", NAME));
+    }
+
+    @Override
+    public void start() {
+        monitor.info(String.format("Started %s extension", NAME));
+    }
+
+    @Override
+    public void shutdown() {
+        monitor.info(String.format("Shutdowns %s extension", NAME));
+    }
+
+    private CosmosDatabase createDatabase(AbstractCosmosConfig configuration, CosmosClient client) {
+        return getDatabase(client, configuration.getDbName());
+    }
+
+    private CosmosClient createClient(Vault vault, String accountName, List<String> preferredRegions) {
+        var accountKey = vault.resolveSecret(accountName);
+        if (StringUtils.isNullOrEmpty(accountKey)) {
+            throw new EdcException("No credentials found in vault for Cosmos DB '" + accountName + "'");
+        }
+
+        // create cosmos db api client
+        String host = "https://" + accountName + ".documents.azure.com:443/";
```

Comment:
we might want to convert this into a constant and format it here with `String.format`

Reply:
done